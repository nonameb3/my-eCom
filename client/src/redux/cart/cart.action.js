import * as TYPE from "./cart.type";

export const toggleCart = () => {
  return {
    type: TYPE.TOGGLE_CART_DROPDOWN
  };
};

export const addCartItem = item => {
  return {
    type: TYPE.ADD_CART_ITEMS,
    payload: item
  };
};

export const deleteCartItem = item => {
  return {
    type: TYPE.DELETE_CART_ITEM,
    payload: item
  };
};

export const removeCartItem = item => {
  return {
    type: TYPE.REMOVE_CART_ITEM,
    payload: item
  };
};

export const clearAllCartItems = () => {
  return { 
    type: TYPE.CLEAR_CART
  }
};
