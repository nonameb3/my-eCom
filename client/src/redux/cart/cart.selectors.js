import { createSelector } from "reselect";

// reselect will select only value change
const selectCart = state => state.cart;

export const selecterCartItems = createSelector(
  [selectCart],
  cart => cart.cartItems
);

export const selectCartHidden = createSelector(
  [selectCart],
  cart => cart.hidden
);

export const selecterCartItemsCount = createSelector(
  [selecterCartItems],
  cartItems => cartItems.reduce((preItem, item) => preItem + item.quantity, 0)
);

export const selecterCartItemsTotalCount = createSelector(
  [selecterCartItems],
  cartItems =>
    cartItems.reduce((preItem, item) => preItem + item.quantity * item.price, 0)
);
