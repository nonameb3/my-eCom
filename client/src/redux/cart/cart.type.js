export const TOGGLE_CART_DROPDOWN = "TOGGLE_CART_DROPDOWN";
export const ADD_CART_ITEMS = "ADD_CART_ITEMS";
export const DELETE_CART_ITEM = "DELETE_CART_ITEM";
export const REMOVE_CART_ITEM = "REMOVE_CART_ITEM";
export const CLEAR_CART = "CLEAR_CART";
