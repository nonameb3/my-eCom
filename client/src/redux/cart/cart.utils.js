export const addNewCartItems = (cartitems = [], newCartItem = {}) => {
  // find existing
  const existingCartItem = cartitems.find(
    cartitem => cartitem.id === newCartItem.id
  );

  // +1 quantity if exist
  if (existingCartItem) {
    // return array
    return cartitems.map(cartitem =>
      cartitem.id === newCartItem.id
        ? { ...cartitem, quantity: cartitem.quantity + 1 }
        : cartitem
    );
  }

  // new item
  return [...cartitems, { ...newCartItem, quantity: 1 }];
};

export const removeCartItem = (cartitems = [], removeCartItem = {}) => {
  // find existing
  const existingCartItem = cartitems.find(
    cartitem => cartitem.id === removeCartItem.id
  );

  // -1 quantity if exist
  if (existingCartItem && existingCartItem.quantity > 1) {
    // return array
    return cartitems.map(cartitem =>
      cartitem.id === removeCartItem.id
        ? { ...cartitem, quantity: cartitem.quantity - 1 }
        : cartitem
    );
  }

  return cartitems;
}
