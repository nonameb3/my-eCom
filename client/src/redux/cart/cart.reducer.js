import * as TYPE from "./cart.type";
import { addNewCartItems, removeCartItem } from './cart.utils'

const INITIAL_STATE = {
  hidden: true,
  cartItems: []
};

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case TYPE.TOGGLE_CART_DROPDOWN:
      return {
        ...state,
        hidden: !state.hidden
      };
    case TYPE.ADD_CART_ITEMS:
      return {
        ...state,
        cartItems: addNewCartItems(state.cartItems, action.payload)
      }
    case TYPE.REMOVE_CART_ITEM:
      return {
        ...state,
        cartItems: removeCartItem(state.cartItems, action.payload)
      }
    case TYPE.DELETE_CART_ITEM:
      return {
        ...state,
        cartItems: state.cartItems.filter(item => item.id !== action.payload.id)
      }
    case TYPE.CLEAR_CART:
      return {
        ...state,
        cartItems: []
      }
    default:
      return state;
  }
}
