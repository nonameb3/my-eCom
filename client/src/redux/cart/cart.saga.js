import { all ,put ,call ,takeLatest } from 'redux-saga/effects';

import { clearAllCartItems } from './cart.action';
import * as USERTYPE from '../user/user.type';

export function* clearCartStore() {
  yield put(clearAllCartItems());
}

// handle function
export function* onSingOutSuccess() {
  yield takeLatest(USERTYPE.SIGN_OUT_SUCCESS, clearCartStore)
}

export function* cartSaga() {
  yield all([call(onSingOutSuccess)])
}
