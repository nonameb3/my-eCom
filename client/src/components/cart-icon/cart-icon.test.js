import React from 'react';
import { shallow} from 'enzyme';

import CartIcon from './cart-icon.component';
import { toggleCart } from "../../redux/cart/cart.action";

let wrapper;
const mockDispatch = jest.fn();
const mockItemCount = 99;

jest.mock("react-redux", () => ({
  useSelector: jest.fn(() => mockItemCount),
  useDispatch: () => mockDispatch
}));

describe('Cart-icon component', () => {
  beforeEach(() => {
    wrapper = shallow(<CartIcon />);
  })

  it('should rander cart-icon component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should action toggleCart when button is clicked', () => {
    wrapper.find('.cart-icon').simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(toggleCart());
  });

  it('should rander item count as expect', () => {
    wrapper.find('.cart-icon').simulate('click');
    expect(wrapper.find('.item-count').text()).toEqual(mockItemCount.toString())
  });
})