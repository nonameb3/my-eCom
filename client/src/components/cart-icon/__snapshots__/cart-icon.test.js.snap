// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart-icon component should rander cart-icon component 1`] = `ShallowWrapper {}`;
